modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Unfair.java

```diff
@@ -0,0 +1,875 @@
+            }
+
+            try {
+                if (!isGloballyLocked) {
+                    globalSync.acquireInterruptibly();
+
+                    isGloballyLocked = true;
+
+                    nextFinish = MAX_TIME + System.nanoTime();
+                }
+            }
+            catch (InterruptedException e) {
+                threadCount.decrementAndGet();
+
+                lock.unlock();
+
+                throw e;
+            }
+
+            reentrantCount.increment();
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean tryLock() {
+            if (reentrantCount.get() > 0) {
+                reentrantCount.increment();
+
+                return true;
+            }
+
+            // tryLock doesn't wait for other threads and increment the threadCount only under the lock.
+            if (lock.tryLock()) {
+                if (isGloballyLocked) {
+                    threadCount.incrementAndGet();
+
+                    reentrantCount.increment();
+
+                    return true;
+                }
+                else {
+                    if (globalSync.tryAcquire()) {
+                        isGloballyLocked = true;
+
+                        threadCount.incrementAndGet();
+
+                        reentrantCount.increment();
+
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean tryLock(long time, TimeUnit unit) throws InterruptedException {
+            assert unit != null;
+
+            if (reentrantCount.get() > 0) {
+                reentrantCount.increment();
+
+                return true;
+            }
+
+            long start = System.nanoTime();
+
+            threadCount.incrementAndGet();
+
+            if (lock.tryLock(time, unit)) {
+                if (isGloballyLocked) {
+                    reentrantCount.increment();
+
+                    return true;
+                }
+                else {
+                    long left = unit.toNanos(time) - (System.nanoTime() - start);
+
+                    try {
+                        if (globalSync.acquire(left, TimeUnit.NANOSECONDS)) {
+                            isGloballyLocked = true;
+
+                            reentrantCount.increment();
+
+                            return true;
+                        }
+                    }
+                    catch (InterruptedException e) {
+                        threadCount.decrementAndGet();
+
+                        lock.unlock();
+
+                        throw e;
+                    }
+                }
+            }
+
+            releaseGlobalLockIfNeed();
+
+            return false;
+        }
+
+        /** {@inheritDoc} */
+        @Override public Condition newCondition() {
+            throw new UnsupportedOperationException("IgniteLock does not allow creation of nameless conditions. ");
```

Comment:
mb, throw unsupported exception here and there

Reply:
I take it from the old lock.